tests/contextwatchdog.js

```diff
@@ -0,0 +1,588 @@
+
+		it( 'should throw when multiple items with the same name are added', async () => {
+			watchdog = ContextWatchdog.for( Context, {} );
+
+			watchdog.add( {
+				editor1: {
+					type: 'editor',
+					creator: ( el, config ) => ClassicTestEditor.create( el, config ),
+					sourceElementOrData: element1,
+					config: {}
+				},
+			} );
+
+			await watchdog.waitForReady();
+
+			let err;
+			try {
+				await watchdog.add( {
+					editor1: {
+						type: 'editor',
+						creator: ( el, config ) => ClassicTestEditor.create( el, config ),
+						sourceElementOrData: element1,
+						config: {}
+					},
+				} );
+			} catch ( _err ) {
+				err = _err;
+			}
+
+			await watchdog.destroy();
+
+			expect( err ).to.be.instanceOf( Error );
+			expect( err.message ).to.match( /Item with the given name is already added: 'editor1'./ );
+		} );
+
+		it( 'should throw when not added item is removed', async () => {
+			watchdog = ContextWatchdog.for( Context, {} );
+
+			await watchdog.waitForReady();
+
+			let err;
+
+			try {
+				await watchdog.remove( [ 'foo' ] );
+			} catch ( _err ) {
+				err = _err;
+			}
+
+			await watchdog.destroy();
+
+			expect( err ).to.be.instanceOf( Error );
+			expect( err.message ).to.match( /Item with the given name was not registered: foo\./ );
+		} );
+
+		it( 'should throw when the item is added before the context is created', async () => {
+			const mainWatchdog = new ContextWatchdog( {}, {} );
+
+			let err;
+			try {
+				await mainWatchdog.add( {} );
+			} catch ( _err ) {
+				err = _err;
+			}
+
+			expect( err ).to.be.instanceOf( Error );
+			expect( err.message ).to.match(
+				/Context was not created yet\. You should call the `ContextWatchdog#create\(\)` method first\./
+			);
+		} );
+
+		it( 'should allow setting editor custom destructors', async () => {
+			watchdog = ContextWatchdog.for( Context, {} );
+
+			const destructorSpy = sinon.spy( editor => editor.destroy() );
+
+			watchdog.add( {
+				editor1: {
+					type: 'editor',
+					creator: ( el, config ) => ClassicTestEditor.create( el, config ),
+					destructor: destructorSpy,
+					sourceElementOrData: element1,
+					config: {},
+				},
+			} );
+
+			await watchdog.destroy();
+
+			sinon.assert.calledOnce( destructorSpy );
+		} );
+
+		it( 'should throw when the item is of not known type', async () => {
+			watchdog = ContextWatchdog.for( Context, {} );
+
+			await watchdog.waitForReady();
+
+			let err;
+			try {
+				await watchdog.add( {
+					editor1: {
+						type: 'foo',
+						creator: ( el, config ) => ClassicTestEditor.create( el, config ),
+						sourceElementOrData: element1,
+						config: {}
+					},
+				} );
+			} catch ( _err ) {
+				watchdog._stopErrorHandling();
+				err = _err;
+			}
+
+			await watchdog.destroy();
+
+			expect( err ).to.be.instanceOf( Error );
+			expect( err.message ).to.match( /Not supported item type: 'foo'\./ );
+		} );
+
+		it( 'should allow adding and removing items without waiting', async () => {
```

Comment:
Isn't it similar to `should allow adding multiple items without waiting for promises` test? Except of removing? Also, what exactly "without waiting" means? I can see `waitForReady()` usage. Maybe you could add a single line comment?

Reply:
I've fixed it.

Reply:
I've fixed the test name and the test content.